test/apps/iodemo/ucx_wrapper.cc

```diff
@@ -0,0 +1,591 @@
+/*
+ * Copyright (C) Mellanox Technologies Ltd. 2020.  ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#include "ucx_wrapper.h"
+
+#include <netinet/in.h>
+#include <arpa/inet.h>
+#include <string.h>
+#include <assert.h>
+
+#include <iostream>
+#include <limits>
+#include <sstream>
+
+
+struct ucx_request {
+    UcxCallback  *callback;
+    bool         completed;
+    uint32_t     conn_id;
+    ucs_status_t status;
+};
+
+class debug {
+public:
+    template<typename T>
+    debug& operator<<(const T &t) {
+#if _DEBUG
+        _oss << t;
+#endif
+        return *this; 
+    }
+
+    typedef std::basic_ostream<char, std::char_traits<char> > EndLine;
+    typedef EndLine& (*EndLineManupulator)(EndLine&);
+    debug& operator<<(EndLineManupulator manupulator) {
+#if _DEBUG
+        manupulator(_oss);
+#endif
+        return *this;
+    }
+
+    debug() {
+#if _DEBUG
+        _oss << "[DEBUG] ";
+#endif
+    }
+
+    ~debug() {
+#if _DEBUG
+        std::cout << _oss.str();
```

Comment:
why need to collect all to stringstream and then put to cout? can't just put to cout?

Reply:
do not mess output for multi threading for complex messages like
```
debug() << "this" << " is" << " the" << " long" << " message" << std::endl;
```

Reply:
not sure it's worth it here, nothing else in the code is thread-safe

Reply:
I removed singleton implementation because of your concern on thread safety :)

Reply:
the main goal to remove singleton is to make it simpler..

Reply:
the goal of debug message is don't add overhead for release build, the series of `operator <<` calls with runtime check may impact performance...
Adding
```
        if (opts().verbose) {
            std::cout << ...  << std::endl;
        }
```
for every useful debug/verbose print looks ugly
(macro is not a C++ way)